src/ldapclient_lib.py

```diff
@@ -0,0 +1,395 @@
+        )
+        self.framework.observe(
+            charm.on[integration_name].relation_joined,
+            self._on_relation_joined,
+        )
+        self.charm = charm
+        self.integration_name = integration_name
+
+    def _get_hostname(self) -> str:
+        """Get GLAuth hostname.
+
+        Returns:
+            GLAuth hostname.
+        """
+        hostname = subprocess.run(
+            ["cat", "/etc/hostname"], capture_output=True, text=True
+        ).stdout.strip()
+        return hostname
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle relation-broken event.
+
+        When the ldapclient relation is broken and emits:
+        - Server unavailable event: When the ldap server can't be reached.
+        """
+        # Remove Obsolete Secrets
+        ca_cert_secret = self.model.get_secret(label="ca-cert")
+        ldbd_secret = self.model.get_secret(label="ldap-default-bind-dn")
+        lp_secret = self.model.get_secret(label="ldap-password")
+        ca_cert_secret.remove_all_revisions()
+        ldbd_secret.remove_all_revisions()
+        lp_secret.remove_all_revisions()
+        self.on.server_unavailable.emit()
+
+    def _on_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Event emitted when the relation is joined.
+
+        Looks at the relation data and config values and emits:
+        - config unavailable event: If the config resource is not supplied.
+        - glauth snap event: When the glauth certificate, config, and key are available.
+        """
+        self.charm.unit.status = MaintenanceStatus("reconfiguring glauth")
+        # Check model for GLAuth config resource
+        try:
+            resource_path = self.model.resources.fetch("config")
+        except ModelError:
+            logger.debug("No config resource supplied")
+            resource_path = None
+            self.on.config_data_unavailable.emit(
+                api_port=self.model.config["api-port"], ldap_port=self.model.config["ldap-port"]
+            )
+        # Set config and get LDAP URI
+        ldap_uri = self.set_config(
+            resource_path, self.model.config["ldap-port"], self.model.config["api-port"]
+        )
+        # Get CA Cert and key
+        ca_cert = self.load()
+        # Signals glauth snap is ready to be started
+        self.on.glauth_snap_ready.emit()
+        cc_content = {"ca-cert": ca_cert}
+        # Get Peer Secrets
+        ldap_relation = self.model.get_relation("glauth")
+        default_bind_dn = ldap_relation.data[self.charm.app]["ldap-default-bind-dn"]
+        ldap_password = ldap_relation.data[self.charm.app]["ldap-password"]
+        ldbd_secret = self.model.get_secret(id=default_bind_dn)
+        lp_secret = self.model.get_secret(id=ldap_password)
+        # Create Secrets
+        cc_secret = self.charm.app.add_secret(cc_content, label="ca-cert")
+        logger.debug("created secret %s", cc_secret)
```

Comment:
Just use the f-string :laughing:. We don't actually gain that much performance from using the logger's built-in formatter compared to f-strings, especially when working with primitives like strings and integers: 

```suggestion
        logger.debug(f"created secret {cc_secret}")
```